File: packages/dfos-web-relay-go/signing.go
package relay

import (
	"bytes"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	dfos "github.com/metalabel/dfos/packages/dfos-protocol-go"
)

const (
	maxSigningDepositBody  = 524288
	maxSigningResponse     = 8192
	maxSigningResponseBody = maxSigningResponse + 512
	maxSigningDeclineBody  = 512
)

type bundledSigningIdentity struct {
	state dfos.IdentityState
	log   []string
}

func (r *Relay) signingMailboxStore() SigningStore {
	return r.readStore.(SigningStore)
}

func verifySigningBundle(tokens []string) (map[string]bundledSigningIdentity, error) {
	grouped := make(map[string][]string)
	for _, token := range tokens {
		header, payload, err := dfos.DecodeJWSUnsafe(token)
		if err != nil || header == nil {
			return nil, fmt.Errorf("invalid identity operation in bundle")
		}
		var did string
		if typ, _ := payload["type"].(string); typ == "create" {
			result, err := dfos.VerifyIdentityChain([]string{token})
			if err != nil {
				return nil, err
			}
			did = result.State.DID
		} else {
			hash := strings.Index(header.Kid, "#")
			if hash < 0 {
				return nil, fmt.Errorf("invalid identity operation kid in bundle")
			}
			did = header.Kid[:hash]
		}
		grouped[did] = append(grouped[did], token)
	}

	verified := make(map[string]bundledSigningIdentity)
	for did, log := range grouped {
		result, err := dfos.VerifyIdentityChain(log)
		if err != nil {
			return nil, err
		}
		if result.State.DID != did {
			return nil, fmt.Errorf("identity bundle DID mismatch")
		}
		verified[did] = bundledSigningIdentity{state: result.State, log: log}
	}
	return verified, nil
}

func keyFromState(state dfos.IdentityState, keyID string) (ed25519.PublicKey, error) {
	keys := make([]dfos.MultikeyPublicKey, 0, len(state.AuthKeys)+len(state.AssertKeys)+len(state.ControllerKeys))
	keys = append(keys, state.AuthKeys...)
	keys = append(keys, state.AssertKeys...)
	keys = append(keys, state.ControllerKeys...)
	for _, key := range keys {
		if key.ID == keyID {
			return dfos.DecodeMultikey(key.PublicKeyMultibase)
		}
	}
	return nil, fmt.Errorf("unknown key %s", keyID)
}

func signingBundleKey(identity bundledSigningIdentity, kid string, historical bool) (ed25519.PublicKey, error) {
	hash := strings.Index(kid, "#")
	if hash < 0 || kid[:hash] != identity.state.DID {
		return nil, fmt.Errorf("invalid kid")
	}
	keyID := kid[hash+1:]
	if key, err := keyFromState(identity.state, keyID); err == nil {
		return key, nil
	}
	if historical {
		for _, token := range identity.log {
			_, payload, err := dfos.DecodeJWSUnsafe(token)
			if err != nil {
				continue
			}
			for _, role := range []string{"authKeys", "assertKeys", "controllerKeys"} {
				entries, _ := payload[role].([]any)
				for _, raw := range entries {
					entry, _ := raw.(map[string]any)
					if id, _ := entry["id"].(string); id == keyID {
						multibase, _ := entry["publicKeyMultibase"].(string)
						return dfos.DecodeMultikey(multibase)
					}
				}
			}
		}
	}
	return nil, fmt.Errorf("unknown key %s", keyID)
}

func signingResolvers(store Store, bundle map[string]bundledSigningIdentity) (dfos.KeyResolver, dfos.KeyResolver) {
	localCurrent := CreateCurrentKeyResolver(store)
	localHistorical := CreateKeyResolver(store)
	current := func(kid string) (ed25519.PublicKey, error) {
		hash := strings.Index(kid, "#")
		if hash < 0 {
			return nil, fmt.Errorf("invalid kid")
		}
		if local, _ := store.GetIdentityChain(kid[:hash]); local != nil {
			if local.State.IsDeleted {
				return nil, fmt.Errorf("identity deleted")
			}
			return localCurrent(kid)
		}
		identity, ok := bundle[kid[:hash]]
		if !ok || identity.state.IsDeleted {
			return nil, fmt.Errorf("identity unavailable")
		}
		return signingBundleKey(identity, kid, false)
	}
	historical := func(kid string) (ed25519.PublicKey, error) {
		hash := strings.Index(kid, "#")
		if hash < 0 {
			return nil, fmt.Errorf("invalid kid")
		}
		if local, _ := store.GetIdentityChain(kid[:hash]); local != nil {
			return localHistorical(kid)
		}
		identity, ok := bundle[kid[:hash]]
		if !ok || identity.state.IsDeleted {
			return nil, fmt.Errorf("identity unavailable")
		}
		return signingBundleKey(identity, kid, true)
	}
	return current, historical
}

func mailboxDepositCovered(att []dfos.AttEntry, subject string) bool {
	resource := "mailbox:" + strings.TrimPrefix(subject, "did:dfos:")
	exact := make([]dfos.AttEntry, 0, len(att))
	for _, entry := range att {
		if entry.Resource == resource {
			exact = append(exact, entry)
		}
	}
	return matchesResource(exact, resource, "deposit")
}

func (r *Relay) verifySigningCredential(token string, request *dfos.VerifiedSignRequest, resolveKey dfos.KeyResolver, bundle map[string]bundledSigningIdentity) error {
	header, payload, err := dfos.DecodeJWSUnsafe(token)
	if err != nil || header == nil || !strings.Contains(header.Kid, "#") {
		return fmt.Errorf("invalid credential")
	}
	issuer := header.Kid[:strings.Index(header.Kid, "#")]
	if local, _ := r.readStore.GetIdentityChain(issuer); local != nil && local.State.IsDeleted {
		return fmt.Errorf("issuer deleted")
	}
	if identity, ok := bundle[issuer]; ok && identity.state.IsDeleted {
		return fmt.Errorf("issuer deleted")
	}
	key, err := resolveKey(header.Kid)
	if err != nil {
		return err
	}
	verified, err := dfos.VerifyCredential(token, key, "", "")
	if err != nil {
		return err
	}
	revoked, err := r.readStore.IsCredentialRevoked(verified.Iss, verified.CID, 0)
	if err != nil || revoked {
		return fmt.Errorf("credential revoked")
	}
	att := dfos.ParseAtt(payload)
	prf, err := dfos.ParsePrf(payload)
	if err != nil {
		return err
	}
	isRevoked := func(issuerDID, credentialCID string, _ int64) (bool, error) {
		return r.readStore.IsCredentialRevoked(issuerDID, credentialCID, 0)
	}
	isDeleted := func(did string) (bool, error) {
		if local, err := r.readStore.GetIdentityChain(did); err != nil {
			return false, err
		} else if local != nil {
			return local.State.IsDeleted, nil
		}
		identity, ok := bundle[did]
		return ok && identity.state.IsDeleted, nil
	}
	if err := dfos.VerifyDelegationChain(token, verified, att, prf, resolveKey, request.Subject, isRevoked, isDeleted, 0); err != nil {
		return err
	}
	if verified.Aud != "*" && verified.Aud != request.DID {
		return fmt.Errorf("credential audience mismatch")
	}
	if !mailboxDepositCovered(att, request.Subject) {
		return fmt.Errorf("credential does not cover mailbox deposit")
	}
	return nil
}

func (r *Relay) handleSigningDeposit(w http.ResponseWriter, req *http.Request) {
	if !r.signingEnabled {
		writeError(w, http.StatusNotImplemented, "signing mailbox not available")
		return
	}
	if req.ContentLength > maxSigningDepositBody {
		writeError(w, http.StatusRequestEntityTooLarge, "request body too large")
		return
	}
	req.Body = http.MaxBytesReader(w, req.Body, maxSigningDepositBody)
	var body struct {
		Request    string   `json:"request"`
		Credential string   `json:"credential"`
		Chain      []string `json:"chain"`
	}
	decoder := json.NewDecoder(req.Body)
	if err := decoder.Decode(&body); err != nil {
		var tooLarge *http.MaxBytesError
		if errors.As(err, &tooLarge) {
			writeError(w, http.StatusRequestEntityTooLarge, "request body too large")
		} else {
			writeError(w, http.StatusBadRequest, "invalid JSON body")
		}
		return
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF || body.Request == "" || body.Credential == "" {
		writeError(w, http.StatusBadRequest, "invalid request")
		return
	}
	bundle, err := verifySigningBundle(body.Chain)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid identity chain bundle")
		return
	}
	current, historical := signingResolvers(r.readStore, bundle)
	verified, err := dfos.VerifySignRequest(body.Request, current, time.Now())
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid sign request")
		return
	}
	subject, err := r.readStore.GetIdentityChain(verified.Subject)
	if storeErr(w, err) {
		return
	}
	if subject == nil || subject.State.IsDeleted {
		writeError(w, http.StatusNotFound, "subject identity not found")
		return
	}
	if err := r.verifySigningCredential(body.Credential, verified, historical, bundle); err != nil {
		writeError(w, http.StatusForbidden, "deposit credential rejected")
		return
	}
	now := time.Now().UTC()
	result, err := r.signingMailboxStore().PutSignRequest(StoredSignRequest{
		CID: verified.RequestCID, Request: body.Request, RequesterDID: verified.DID,
		SubjectDID: verified.Subject, PayloadTyp: verified.PayloadTyp,
		PayloadBytes: append([]byte(nil), verified.PayloadBytes...), ExpiresAt: verified.ExpiresAt,
		DepositedAt: now.Format(signingTimeFormat), Declined: false,
	}, now)
	if storeErr(w, err) {
		return
	}
	if result == SigningConflict {
		writeError(w, http.StatusConflict, "request CID conflict")
		return
	}
	if result == SigningAtCapacity {
		writeError(w, http.StatusTooManyRequests, "signing mailbox pending limit reached")
		return
	}
	status := http.StatusCreated
	if result == SigningIdentical {
		status = http.StatusOK
	}
	writeJSON(w, status, map[string]string{"cid": verified.RequestCID, "expiresAt": verified.ExpiresAt})
}

func (r *Relay) handleSigningPoll(w http.ResponseWriter, req *http.Request) {
	if !r.signingEnabled {
		writeError(w, http.StatusNotImplemented, "signing mailbox not available")
		return
	}
	auth := AuthenticateRequest(req.Header.Get("Authorization"), r.did, r.readStore, r.maxAuthTokenTTL)
	if auth == nil {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}
	after := req.URL.Query().Get("after")
	limit := parseLimit(req, 100, 1000)
	requests, nextCursor, err := r.signingMailboxStore().ListPendingSignRequests(auth.Iss, after, limit, time.Now())
	if errors.Is(err, ErrInvalidSigningCursor) {
		writeError(w, http.StatusBadRequest, "invalid cursor")
		return
	}
	if storeErr(w, err) {
		return
	}
	type item struct {
		CID         string `json:"cid"`
		Request     string `json:"request"`
		DepositedAt string `json:"depositedAt"`
		Declined    bool   `json:"declined"`
	}
	items := make([]item, len(requests))
	for i, request := range requests {
		items[i] = item{request.CID, request.Request, request.DepositedAt, request.Declined}
	}
	var next any
	if nextCursor != "" {
		next = nextCursor
	}
	writeJSON(w, http.StatusOK, map[string]any{"requests": items, "next": next})
}

func decodeCanonicalSigningResponse(token string) (*dfos.JWSHeader, []byte, error) {
	if len(token) > maxSigningResponse {
		return nil, nil, fmt.Errorf("response too large")
	}
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, nil, fmt.Errorf("invalid JWS")
	}
	decoded := make([][]byte, len(parts))
	for i, part := range parts {
		segment, err := base64.RawURLEncoding.DecodeString(part)
		if err != nil || base64.RawURLEncoding.EncodeToString(segment) != part {
			return nil, nil, fmt.Errorf("non-canonical JWS segment")
		}
		decoded[i] = segment
	}
	var header dfos.JWSHeader
	if err := json.Unmarshal(decoded[0], &header); err != nil {
		return nil, nil, err
	}
	if header.Typ == "" || header.Kid == "" {
		return nil, nil, fmt.Errorf("invalid response header")
	}
	return &header, decoded[1], nil
}

func (r *Relay) verifySigningResponse(token string, request *StoredSignRequest) error {
	header, payload, err := decodeCanonicalSigningResponse(token)
	if err != nil {
		return err
	}
	if header.Typ != request.PayloadTyp {
		return fmt.Errorf("response typ mismatch")
	}
	hash := strings.Index(header.Kid, "#")
	if hash < 0 || header.Kid[:hash] != request.SubjectDID {
		return fmt.Errorf("response subject mismatch")
	}
	if !bytes.Equal(payload, request.PayloadBytes) {
		return fmt.Errorf("response payload mismatch")
	}
	key, err := CreateKeyResolver(r.readStore)(header.Kid)
	if err != nil {
		return err
	}
	_, _, err = dfos.VerifyJWS(token, key)
	return err
}

func (r *Relay) handleSigningRespond(w http.ResponseWriter, req *http.Request) {
	if !r.signingEnabled {
		writeError(w, http.StatusNotImplemented, "signing mailbox not available")
		return
	}
	now := time.Now()
	request, err := r.signingMailboxStore().GetSignRequest(req.PathValue("cid"), now)
	if storeErr(w, err) {
		return
	}
	if request == nil {
		writeError(w, http.StatusNotFound, "sign request not found")
		return
	}
	req.Body = http.MaxBytesReader(w, req.Body, maxSigningResponseBody)
	var body struct {
		Response string `json:"response"`
	}
	decoder := json.NewDecoder(req.Body)
	if err := decoder.Decode(&body); err != nil {
		var tooLarge *http.MaxBytesError
		if errors.As(err, &tooLarge) {
			writeError(w, http.StatusRequestEntityTooLarge, "request body too large")
		} else {
			writeError(w, http.StatusBadRequest, "invalid response")
		}
		return
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		var tooLarge *http.MaxBytesError
		if errors.As(err, &tooLarge) {
			writeError(w, http.StatusRequestEntityTooLarge, "request body too large")
		} else {
			writeError(w, http.StatusBadRequest, "invalid response")
		}
		return
	}
	if body.Response == "" {
		writeError(w, http.StatusBadRequest, "invalid response")
		return
	}
	if err := r.verifySigningResponse(body.Response, request); err != nil {
		writeError(w, http.StatusBadRequest, "invalid response")
		return
	}
	result, err := r.signingMailboxStore().PutSignResponse(request.CID, body.Response, time.Now())
	if storeErr(w, err) {
		return
	}
	if result == SigningNotFound {
		writeError(w, http.StatusNotFound, "sign request not found")
		return
	}
	if result == SigningConflict {
		writeError(w, http.StatusConflict, "response already exists")
		return
	}
	status := http.StatusCreated
	if result == SigningIdentical {
		status = http.StatusOK
	}
	writeJSON(w, status, map[string]string{"status": "stored"})
}

func (r *Relay) handleSigningGetResponse(w http.ResponseWriter, req *http.Request) {
	if !r.signingEnabled {
		writeError(w, http.StatusNotImplemented, "signing mailbox not available")
		return
	}
	request, err := r.signingMailboxStore().GetSignRequest(req.PathValue("cid"), time.Now())
	if storeErr(w, err) {
		return
	}
	if request == nil {
		writeError(w, http.StatusNotFound, "sign request not found")
		return
	}
	if request.Response != "" {
		writeJSON(w, http.StatusOK, map[string]string{"status": "responded", "response": request.Response})
		return
	}
	if request.Declined {
		writeJSON(w, http.StatusOK, map[string]string{"status": "declined"})
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "pending"})
}

func (r *Relay) handleSigningDecline(w http.ResponseWriter, req *http.Request) {
	if !r.signingEnabled {
		writeError(w, http.StatusNotImplemented, "signing mailbox not available")
		return
	}
	req.Body = http.MaxBytesReader(w, req.Body, maxSigningDeclineBody)
	body, err := io.ReadAll(req.Body)
	if err != nil {
		var tooLarge *http.MaxBytesError
		if errors.As(err, &tooLarge) {
			writeError(w, http.StatusRequestEntityTooLarge, "request body too large")
		} else {
			writeError(w, http.StatusBadRequest, "invalid request body")
		}
		return
	}
	if len(bytes.TrimSpace(body)) != 0 {
		writeError(w, http.StatusBadRequest, "request body must be empty")
		return
	}
	result, err := r.signingMailboxStore().DeclineSignRequest(req.PathValue("cid"), time.Now())
	if storeErr(w, err) {
		return
	}
	if result == SigningNotFound {
		writeError(w, http.StatusNotFound, "sign request not found")
		return
	}
	if result == SigningConflict {
		writeError(w, http.StatusConflict, "response already exists")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

File: packages/dfos-web-relay/src/signing.ts
/* Optional SIGNING 0.1 mailbox courier routes. */

import {
  decodeMultikey,
  verifyIdentityChain,
  verifySignRequest,
  type VerifiedIdentity,
} from '@metalabel/dfos-protocol/chain';
import {
  matchesResource,
  verifyDelegationChain,
  verifyDFOSCredential,
} from '@metalabel/dfos-protocol/credentials';
import {
  assertJwsProfile,
  base64urlDecode,
  base64urlEncode,
  decodeJwsUnsafe,
  isValidEd25519Signature,
} from '@metalabel/dfos-protocol/crypto';
import { Hono, type Context } from 'hono';
import { authenticateRequest, DEFAULT_MAX_AUTH_TOKEN_TTL_SECONDS } from './auth';
import { createHistoricalIdentityResolver, mergeHistoricalIdentity } from './ingest';
import { decodeSigningCursor } from './types';
import type { RelayStore, SigningStore, StoredSignRequest } from './types';

const MAX_DEPOSIT_BODY_BYTES = 524_288;
const MAX_RESPONSE_TOKEN_BYTES = 8_192;
const MAX_RESPONSE_BODY_BYTES = MAX_RESPONSE_TOKEN_BYTES + 512;
const MAX_DECLINE_BODY_BYTES = 512;
const encoder = new TextEncoder();
const decoder = new TextDecoder();

const parseSigningLimit = (raw: string | undefined): number => {
  if (raw === undefined || raw === '') return 100;
  if (!/^-?\d+$/.test(raw)) return 100;
  const limit = Number(raw);
  if (!Number.isSafeInteger(limit) || limit < 1) return 100;
  return Math.min(limit, 1000);
};

type BundledIdentity = { state: VerifiedIdentity; log: string[] };

/** Verify every supplied identity chain and index it ephemerally by DID. */
const verifyBundle = async (tokens: string[]): Promise<Map<string, BundledIdentity>> => {
  const grouped = new Map<string, string[]>();
  for (const token of tokens) {
    const decoded = decodeJwsUnsafe(token);
    if (!decoded) throw new Error('invalid identity operation in chain bundle');
    const payload = decoded.payload as Record<string, unknown>;
    let did: string;
    if (payload['type'] === 'create') {
      did = (await verifyIdentityChain({ didPrefix: 'did:dfos', log: [token] })).did;
    } else {
      const kid = decoded.header.kid;
      const hash = typeof kid === 'string' ? kid.indexOf('#') : -1;
      if (hash < 0) throw new Error('invalid identity operation kid in chain bundle');
      did = kid.substring(0, hash);
    }
    grouped.set(did, [...(grouped.get(did) ?? []), token]);
  }

  const verified = new Map<string, BundledIdentity>();
  for (const [did, log] of grouped) {
    const state = await verifyIdentityChain({ didPrefix: 'did:dfos', log });
    if (state.did !== did) throw new Error('identity bundle DID mismatch');
    verified.set(did, { state, log });
  }
  return verified;
};

const bytesEqual = (a: Uint8Array, b: Uint8Array): boolean =>
  a.length === b.length && a.every((byte, index) => byte === b[index]);

const exactMailboxDeposit = async (
  att: { resource: string; action: string }[],
  subjectDID: string,
): Promise<boolean> => {
  const resource = `mailbox:${subjectDID.replace(/^did:dfos:/, '')}`;
  const exactEntries = att.filter((entry) => entry.resource === resource);
  return matchesResource(exactEntries, resource, 'deposit');
};

const readCappedBytes = async (
  request: Request,
  maximum: number,
): Promise<{ ok: true; bytes: Uint8Array } | { ok: false }> => {
  const declared = Number(request.headers.get('content-length'));
  if (Number.isFinite(declared) && declared > maximum) {
    return { ok: false };
  }
  // With no Content-Length, serve.ts still bounds the incoming stream at 17 MiB;
  // this aggregate check enforces the route-specific ceiling after materialization.
  const bytes = new Uint8Array(await request.arrayBuffer());
  if (bytes.length > maximum) return { ok: false };
  return { ok: true, bytes };
};

const readCappedJSON = async (
  request: Request,
  maximum: number,
): Promise<{ ok: true; body: unknown } | { ok: false; tooLarge: boolean }> => {
  const read = await readCappedBytes(request, maximum);
  if (!read.ok) return { ok: false, tooLarge: true };
  try {
    return { ok: true, body: JSON.parse(decoder.decode(read.bytes)) };
  } catch {
    return { ok: false, tooLarge: false };
  }
};

const verifyResponse = async (response: string, request: StoredSignRequest, store: RelayStore) => {
  if (encoder.encode(response).length > MAX_RESPONSE_TOKEN_BYTES) {
    throw new Error('response exceeds maximum size');
  }
  const parts = response.split('.');
  if (parts.length !== 3) throw new Error('invalid response JWS');
  let header: Record<string, unknown>;
  try {
    const headerBytes = base64urlDecode(parts[0]!);
    if (base64urlEncode(headerBytes) !== parts[0]) throw new Error('non-canonical header');
    header = JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(headerBytes)) as Record<
      string,
      unknown
    >;
  } catch {
    throw new Error('invalid response JWS');
  }
  assertJwsProfile(header, (message) => new Error(message));
  if (header['typ'] !== request.payloadTyp) throw new Error('response typ mismatch');
  const kid = header['kid'];
  if (typeof kid !== 'string') throw new Error('response kid missing');
  const hash = kid.indexOf('#');
  if (hash < 0 || kid.substring(0, hash) !== request.subjectDID) {
    throw new Error('response signer does not match subject');
  }
  const payloadBytes = base64urlDecode(parts[1]!);
  if (base64urlEncode(payloadBytes) !== parts[1]) throw new Error('invalid response payload');
  if (!bytesEqual(payloadBytes, request.payloadBytes)) throw new Error('response payload mismatch');

  const resolveIdentity = createHistoricalIdentityResolver(store);
  const identity = await resolveIdentity(request.subjectDID);
  if (!identity) throw new Error('subject identity unavailable');
  const keyID = kid.substring(hash + 1);
  const key = [...identity.authKeys, ...identity.assertKeys, ...identity.controllerKeys].find(
    (candidate) => candidate.id === keyID,
  );
  if (!key) throw new Error('response key unavailable');
  const signature = base64urlDecode(parts[2]!);
  if (base64urlEncode(signature) !== parts[2]) throw new Error('invalid response signature');
  const signingInput = encoder.encode(`${parts[0]}.${parts[1]}`);
  if (
    !isValidEd25519Signature(
      signingInput,
      signature,
      decodeMultikey(key.publicKeyMultibase).keyBytes,
    )
  ) {
    throw new Error('invalid response signature');
  }
};

export const registerSigningRoutes = (options: {
  app: Hono;
  store: RelayStore;
  relayDID: string;
  basePath: string;
  enabled: boolean;
  maxAuthTokenTTLSeconds?: number;
}): void => {
  const { app, store, relayDID, basePath, enabled } = options;
  const signingStore = store as SigningStore;
  const unavailable = (c: Context) => c.json({ error: 'signing mailbox not available' }, 501);

  app.post(`${basePath}/requests`, async (c) => {
    if (!enabled) return unavailable(c);
    const read = await readCappedJSON(c.req.raw, MAX_DEPOSIT_BODY_BYTES);
    if (!read.ok) {
      return c.json(
        { error: read.tooLarge ? 'request body too large' : 'invalid JSON body' },
        read.tooLarge ? 413 : 400,
      );
    }
    const body = read.body as Record<string, unknown> | null;
    if (
      !body ||
      typeof body !== 'object' ||
      typeof body.request !== 'string' ||
      typeof body.credential !== 'string' ||
      (body.chain !== undefined &&
        (!Array.isArray(body.chain) || !body.chain.every((token) => typeof token === 'string')))
    ) {
      return c.json({ error: 'invalid request' }, 400);
    }

    let bundle: Map<string, BundledIdentity>;
    try {
      bundle = await verifyBundle((body.chain as string[] | undefined) ?? []);
    } catch {
      return c.json({ error: 'invalid identity chain bundle' }, 400);
    }

    const resolveCurrent = async (did: string): Promise<VerifiedIdentity | undefined> => {
      const local = await store.getIdentityChain(did);
      return local?.state ?? bundle.get(did)?.state;
    };
    const localHistorical = createHistoricalIdentityResolver(store);
    const resolveHistorical = async (did: string): Promise<VerifiedIdentity | undefined> => {
      const local = await localHistorical(did);
      const bundled = bundle.get(did);
      return local ?? (bundled ? mergeHistoricalIdentity(bundled.state, bundled.log) : undefined);
    };

    let verifiedRequest: Awaited<ReturnType<typeof verifySignRequest>>;
    try {
      verifiedRequest = await verifySignRequest(body.request, { resolveIdentity: resolveCurrent });
    } catch {
      return c.json({ error: 'invalid sign request' }, 400);
    }

    const subject = await store.getIdentityChain(verifiedRequest.subject);
    if (!subject || subject.state.isDeleted) {
      return c.json({ error: 'subject identity not found' }, 404);
    }

    try {
      const credential = await verifyDFOSCredential(body.credential, {
        resolveIdentity: resolveHistorical,
      });
      if (await store.isCredentialRevoked(credential.iss, credential.credentialCID)) {
        throw new Error('credential revoked');
      }
      await verifyDelegationChain(credential, {
        resolveIdentity: resolveHistorical,
        rootDID: verifiedRequest.subject,
        isRevoked: (issuer, cid) => store.isCredentialRevoked(issuer, cid),
      });
      if (credential.aud !== '*' && credential.aud !== verifiedRequest.did) {
        throw new Error('credential audience mismatch');
      }
      if (!(await exactMailboxDeposit(credential.att, verifiedRequest.subject))) {
        throw new Error('credential does not cover mailbox deposit');
      }
    } catch {
      return c.json({ error: 'deposit credential rejected' }, 403);
    }

    const now = Date.now();
    const result = await signingStore.putSignRequest(
      {
        cid: verifiedRequest.requestCID,
        request: body.request,
        requesterDID: verifiedRequest.did,
        subjectDID: verifiedRequest.subject,
        payloadTyp: verifiedRequest.payloadTyp,
        payloadBytes: verifiedRequest.payloadBytes,
        expiresAt: verifiedRequest.expiresAt,
        depositedAt: new Date(now).toISOString(),
        declined: false,
      },
      now,
    );
    if (result === 'conflict') return c.json({ error: 'request CID conflict' }, 409);
    return c.json(
      { cid: verifiedRequest.requestCID, expiresAt: verifiedRequest.expiresAt },
      result === 'created' ? 201 : 200,
    );
  });

  app.get(`${basePath}/requests`, async (c) => {
    if (!enabled) return unavailable(c);
    const auth = await authenticateRequest(
      c.req.header('authorization'),
      relayDID,
      store,
      options.maxAuthTokenTTLSeconds ?? DEFAULT_MAX_AUTH_TOKEN_TTL_SECONDS,
    );
    if (!auth) return c.json({ error: 'authentication required' }, 401);
    const after = c.req.query('after');
    if (after && !decodeSigningCursor(after)) return c.json({ error: 'invalid cursor' }, 400);
    const limit = parseSigningLimit(c.req.query('limit'));
    const result = await signingStore.listPendingSignRequests({
      subjectDID: auth.iss,
      ...(after ? { after } : {}),
      limit,
      now: Date.now(),
    });
    return c.json({
      requests: result.requests.map((request) => ({
        cid: request.cid,
        request: request.request,
        depositedAt: request.depositedAt,
        declined: request.declined,
      })),
      next: result.next,
    });
  });

  app.post(`${basePath}/requests/:cid/response`, async (c) => {
    if (!enabled) return unavailable(c);
    const request = await signingStore.getSignRequest(c.req.param('cid'), Date.now());
    if (!request) return c.json({ error: 'sign request not found' }, 404);
    const read = await readCappedJSON(c.req.raw, MAX_RESPONSE_BODY_BYTES);
    if (!read.ok) {
      return c.json(
        { error: read.tooLarge ? 'request body too large' : 'invalid JSON body' },
        read.tooLarge ? 413 : 400,
      );
    }
    const body = read.body;
    const response = (body as { response?: unknown } | null)?.response;
    if (typeof response !== 'string') return c.json({ error: 'invalid response' }, 400);
    try {
      await verifyResponse(response, request, store);
    } catch {
      return c.json({ error: 'invalid response' }, 400);
    }
    const result = await signingStore.putSignResponse(request.cid, response, Date.now());
    if (result === 'not-found') return c.json({ error: 'sign request not found' }, 404);
    if (result === 'conflict') return c.json({ error: 'response already exists' }, 409);
    return c.json({ status: 'stored' }, result === 'created' ? 201 : 200);
  });

  app.get(`${basePath}/requests/:cid/response`, async (c) => {
    if (!enabled) return unavailable(c);
    const request = await signingStore.getSignRequest(c.req.param('cid'), Date.now());
    if (!request) return c.json({ error: 'sign request not found' }, 404);
    if (request.response !== undefined) {
      return c.json({ status: 'responded', response: request.response });
    }
    return c.json({ status: request.declined ? 'declined' : 'pending' });
  });

  app.post(`${basePath}/requests/:cid/decline`, async (c) => {
    if (!enabled) return unavailable(c);
    const read = await readCappedBytes(c.req.raw, MAX_DECLINE_BODY_BYTES);
    if (!read.ok) return c.json({ error: 'request body too large' }, 413);
    if (read.bytes.some((byte) => ![9, 10, 13, 32].includes(byte))) {
      return c.json({ error: 'request body must be empty' }, 400);
    }
    const result = await signingStore.declineSignRequest(c.req.param('cid'), Date.now());
    if (result === 'not-found') return c.json({ error: 'sign request not found' }, 404);
    if (result === 'responded') return c.json({ error: 'response already exists' }, 409);
    return c.body(null, 204);
  });
};
